File: android/src/main/java/com/jjlf/rngradient/LinearGradientViewManager.java
package com.jjlf.rngradient;

import android.graphics.Color;
import android.graphics.Shader;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.ViewProps;
import com.facebook.react.uimanager.annotations.ReactProp;



public class LinearGradientViewManager extends ViewGroupManager<LinearGradientView> {
    

    @Override
    public String getName() { return "LinearGradient"; }


    @Override
    protected LinearGradientView createViewInstance(@NonNull ThemedReactContext reactContext) {
        return new LinearGradientView(reactContext);
    }

    @Override
    @ReactProp(
            name = ViewProps.BACKGROUND_COLOR,
            defaultInt = Color.TRANSPARENT,
            customType = "Color")
    public void setBackgroundColor(@NonNull LinearGradientView view, int backgroundColor) {
        
    }

    @ReactProp(name = "startPoint")
    public void setStartPoint(LinearGradientView view, ReadableMap m){
        float x = (float) ModUtil.getDouble(m,"x",0.5);
        float y = (float) ModUtil.getDouble(m,"y",0.0);
        view.getDrawable().setStartX(x);
        view.getDrawable().setStartY(y);
        view.getDrawable().invalidateSelf();

    }

    @ReactProp(name = "endPoint")
    public void setEndPoint(LinearGradientView view, ReadableMap m){
        float x = (float) ModUtil.getDouble(m,"x",0.5);
        float y = (float) ModUtil.getDouble(m,"y",1.0);
        view.getDrawable().setEndX(x);
        view.getDrawable().setEndY(y);
        view.getDrawable().invalidateSelf();
    }

    @ReactProp(name = "tileMode")
    public void setTileMode(LinearGradientView view, String v){
        Shader.TileMode mode = "mirror".equals(v) ? Shader.TileMode.MIRROR
                : ( "repeat".equals(v) ? Shader.TileMode.REPEAT : Shader.TileMode.CLAMP );
        view.getDrawable().setTileMode(mode);
        view.getDrawable().invalidateSelf();
    }
    @ReactProp(name = "colors")
    public void setColors(LinearGradientView view, ReadableArray colors)  {
        int[] c = colors != null ? ModUtil.toIntArray(colors) : new int[] { Color.WHITE, Color.BLACK};
        view.getDrawable().setColors(c);
        view.getDrawable().invalidateSelf();
    }
    @ReactProp(name = "positions")
    public void setPositions(LinearGradientView view, ReadableArray positions)  {
        float[] p = positions != null ? ModUtil.toFloatArray(positions) : null;
        view.getDrawable().setPositions(p);
        view.getDrawable().invalidateSelf();
    }
}

File: android/src/main/java/com/jjlf/rngradient/LinearGradientViewManager.kt
package com.jjlf.rngradient

import android.content.Context
import android.graphics.Color
import android.graphics.Shader
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.uimanager.*


class LinearGradientViewManager : ViewGroupManager<LinearGradientView>() {

    override fun getName(): String {
        return "LinearGradient"
    }

    override fun createViewInstance(reactContext: ThemedReactContext): LinearGradientView {
        return LinearGradientView(reactContext as Context)
    }
    @Override
    @ReactProp(
        name = ViewProps.BACKGROUND_COLOR,
        defaultInt = Color.TRANSPARENT,
        customType = "Color")
    override fun setBackgroundColor(view: LinearGradientView, c:Int){}

    @ReactProp(name = "startPoint")
     fun setStartPoint(view: LinearGradientView, m:ReadableMap?){
        val x = try{ m!!.getDouble("x") }catch (e:Error){ 0.5 }.toFloat()
        val y = try{ m!!.getDouble("y") }catch (e:Error){ 0.0 }.toFloat()
         val mDrawable = view.getDrawable()
        mDrawable.setStartX(x)
        mDrawable.setStartY(y)
        mDrawable.invalidateSelf()

    }

    @ReactProp(name = "endPoint",)
    fun setEndPoint(view: LinearGradientView, m:ReadableMap?){
        val x = try{ m!!.getDouble("x") }catch (e:Error){ 0.5 }.toFloat()
        val y = try{ m!!.getDouble("y") }catch (e:Error){ 1.0 }.toFloat()
        val mDrawable = view.getDrawable()
        mDrawable.setEndX(x)
        mDrawable.setEndY(y)
        mDrawable.invalidateSelf()
    }

    @ReactProp(name = "tileMode")
    fun setTileMode(view: LinearGradientView, v:String?){
        val mDrawable = view.getDrawable()
        val mode = when(v ?: "clamp"){
            "mirror" -> Shader.TileMode.MIRROR
            "repeat" -> Shader.TileMode.REPEAT
            else -> Shader.TileMode.CLAMP
        }
        mDrawable.setTileMode(mode)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "colors")
    fun setColors(view: LinearGradientView, colors: ReadableArray?)  {
        val mDrawable = view.getDrawable()
        val c = if(colors != null) GradientHelper.toIntArray(colors) else intArrayOf(
            Color.WHITE,
            Color.BLACK)
        mDrawable.setColors(c)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "positions")
    fun setPositions(view: LinearGradientView, positions: ReadableArray?)  {
        val mDrawable = view.getDrawable()
        val p = if(positions != null) GradientHelper.toFloatArray(positions) else null
        mDrawable.setPositions(p)
        mDrawable.invalidateSelf()
    }

}